src/checked.rs

```diff
@@ -0,0 +1,419 @@
+  type Bits = u8;
+
+  #[inline]
+  fn is_valid_bit_pattern(bits: &Self::Bits) -> bool {
+    matches!(*bits, 0 | 1)
+  }
+}
+
+/// The things that can go wrong when casting between [`CheckedBitPattern`] data forms.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum CheckedCastError {
+  /// An error occurred during a true-[`Pod`] cast
+  PodCastError(crate::PodCastError),
+  /// When casting to a [`CheckedBitPattern`] type, it is possible that the original
+  /// data contains an invalid bit pattern. If so, the cast will fail and
+  /// this error will be returned. Will never happen on casts between
+  /// [`Pod`] types.
+  InvalidBitPattern,
+}
+
+#[cfg(not(target_arch = "spirv"))]
+impl core::fmt::Display for CheckedCastError {
+  fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+    write!(f, "{:?}", self)
+  }
+}
+#[cfg(feature = "extern_crate_std")]
+impl std::error::Error for CheckedCastError {}
+
+impl From<crate::PodCastError> for CheckedCastError {
+  fn from(err: crate::PodCastError) -> CheckedCastError {
+    CheckedCastError::PodCastError(err)
+  }
+}
+
+/// Re-interprets `&[u8]` as `&T`.
+///
+/// ## Failure
+///
+/// * If the slice isn't aligned for the new type
+/// * If the slice's length isn’t exactly the size of the new type
+/// * If the slice contains an invalid bit pattern for `T`
+#[inline]
+pub fn try_from_bytes<T: CheckedBitPattern>(
+  s: &[u8],
+) -> Result<&T, CheckedCastError> {
+  let pod = unsafe { internal::try_from_bytes(s)? };
+
+  if <T as CheckedBitPattern>::is_valid_bit_pattern(pod) {
+    Ok(unsafe { &*(pod as *const <T as CheckedBitPattern>::Bits as *const T) })
+  } else {
+    Err(CheckedCastError::InvalidBitPattern)
+  }
+}
+
+/// Re-interprets `&mut [u8]` as `&mut T`.
+///
+/// ## Failure
+///
+/// * If the slice isn't aligned for the new type
+/// * If the slice's length isn’t exactly the size of the new type
+/// * If the slice contains an invalid bit pattern for `T`
+#[inline]
+pub fn try_from_bytes_mut<T: CheckedBitPattern + NoPadding>(
+  s: &mut [u8],
+) -> Result<&mut T, CheckedCastError> {
+  let pod = unsafe { internal::try_from_bytes_mut(s)? };
+
+  if <T as CheckedBitPattern>::is_valid_bit_pattern(pod) {
+    Ok(unsafe { &mut *(pod as *mut <T as CheckedBitPattern>::Bits as *mut T) })
+  } else {
+    Err(CheckedCastError::InvalidBitPattern)
+  }
+}
+
+/// Try to cast `T` into `U`.
+///
+/// Note that for this particular type of cast, alignment isn't a factor. The
+/// input value is semantically copied into the function and then returned to a
+/// new memory location which will have whatever the required alignment of the
+/// output type is.
+///
+/// ## Failure
+///
+/// * If the types don't have the same size this fails.
+/// * If `a` contains an invalid bit pattern for `B` this fails.
+#[inline]
+pub fn try_cast<A: NoPadding, B: CheckedBitPattern>(
+  a: A,
+) -> Result<B, CheckedCastError> {
+  let pod = unsafe { internal::try_cast(a) }?;
```

Comment:
```suggestion
  let pod = unsafe { internal::try_cast(a)? };
```
I have no reason to prefer either form, but the functions above do it one way and the functions below do it a different way

Reply:
I'd say "Keep least amount of ops in the `unsafe` block".
Therefore `?` should be outside unsafe block, unless unwrapped value is used in the unsafe block later. Which is not the case here.